bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/InterleavedLedgerStorage.java

```diff
@@ -641,4 +649,233 @@ public void onRotateEntryLog() {
+                curSequenceGroupValue.lastSequenceStart = curSequenceStartEntryId.longValue();
+                curSequenceGroupValue.sequencePeriod = ((int) (curSequenceGroupValue.lastSequenceStart
+                        - curSequenceGroupValue.firstSequenceStart));
+            } else {
+                curSequenceGroupValue.lastSequenceStart = curSequenceStartEntryId.longValue();
+            }
+        }
+
+        private void addCurSequenceToSequenceGroup() {
+            if (!isCurSequenceGroupInitialized()) {
+                initializeCurSequenceGroupWithCurSequence();
+            } else if (doesCurSequenceBelongToCurSequenceGroup()) {
+                appendCurSequenceToCurSequenceGroup();
+            } else {
+                createNewSequenceGroupWithCurSequence();
+            }
+        }
+
+        void addEntryToAvailabileEntriesOfLedger(long entryId) {
+            if (!isCurSequenceInitialized()) {
+                initializeCurSequence(entryId);
+            } else if (isEntryExistingInCurSequence(entryId)) {
+                /* this entry is already added so do nothing */
+            } else if (isEntryAppendableToCurSequence(entryId)) {
+                incrementCurSequenceSize();
+            } else {
+                addCurSequenceToSequenceGroup();
+                initializeCurSequence(entryId);
+            }
+        }
+
+        void closeStateOfEntriesOfALedger() {
+            if (isCurSequenceInitialized()) {
+                addCurSequenceToSequenceGroup();
+                resetCurSequence();
+            }
+        }
+
+        byte[] serializeStateOfEntriesOfALedger() {
+            byte[] header = new byte[HEADER_SIZE];
+            ByteBuffer headerByteBuf = ByteBuffer.wrap(header);
+            byte[] serializedSequenceGroupByteArray = new byte[SequenceGroup.SEQUENCEGROUP_BYTES];
+            byte[] serializedStateByteArray = new byte[HEADER_SIZE
+                    + (listOfSequenceGroups.size() * SequenceGroup.SEQUENCEGROUP_BYTES)];
+            final int numOfSequenceGroups = listOfSequenceGroups.size();
+            headerByteBuf.putInt(CURRENT_HEADER_VERSION);
+            headerByteBuf.putInt(numOfSequenceGroups);
+            System.arraycopy(header, 0, serializedStateByteArray, 0, HEADER_SIZE);
+            int seqNum = 0;
+            for (SequenceGroup seqGroup : listOfSequenceGroups) {
+                Arrays.fill(serializedSequenceGroupByteArray, (byte) 0);
+                seqGroup.serializeSequenceGroup(serializedSequenceGroupByteArray);
+                System.arraycopy(serializedSequenceGroupByteArray, 0, serializedStateByteArray,
+                        HEADER_SIZE + ((seqNum++) * SequenceGroup.SEQUENCEGROUP_BYTES),
+                        SequenceGroup.SEQUENCEGROUP_BYTES);
+            }
+            return serializedStateByteArray;
+        }
+    }
+
+    protected byte[] getEntriesOfLedger(long ledgerId, Iterator<EntryKey> entriesInMemtable) throws Exception {
```

Comment:
This shouldn't take entriesInMemtable as an input -- InterleavedLedgerStorage doesn't necessarily require a memtable.  Separate the above representation classes out into their own files with unit tests.  Modify this method to return an Iterator.  SortedMemTable would be responsible for merging this iterator with the one returned from the memtable.  The SequenceGroup above could simply take an interator in its constructor.

Reply:
new logic should take care of this